### ISSUE TITLE
[QUESTION] - How  to connect to remote server using ssh config?

### ISSUE THREAD

**Issue by namespaceYcZ** [veeso/termscp#149]
Some basic background info:
I have some Linux virtual machines, for some reason they need to use dhcp and their ip addresses change often. I wrote a script to detect their ip addresses (using `arp` util) and write to the hosts file of the host machine, so that I don't have to change my ssh config or any other config file that use the services on those virtual machines.
![2023-02-23_203855](https://user-images.githubusercontent.com/3965259/220912586-0fd1304e-2c7f-4460-911e-934e67c32b09.jpg)

I can always log in to these virtual machines with a fixed ssh command just like this(`ssh cross-platform` where "cross-platform" is the hostname in the ssh config file):
![2023-02-23_202140](https://user-images.githubusercontent.com/3965259/220912748-e07f2345-c2c3-43c8-80f8-a59eeede6ce5.jpg)
 (because I have set up my ssh config and added my public key to the authorized_keys file so I do not have to offer my password)
![2023-02-23_201702](https://user-images.githubusercontent.com/3965259/220913615-dc037bcc-0315-43e0-a29b-35074046f0e5.jpg)

(I can always ssh to it after using the script to detect its IP address even if the IP address of this virtual machine changes).
Now, how can I use termscp to connect to the remote server using ssh config?
（I have check the config of termscp's default config, which is ok:
![2023-02-23_203347](https://user-images.githubusercontent.com/3965259/220913902-b79754db-de60-4793-a1f0-42cc5ea6cfcd.jpg)）

1. using hostname in tui:
![2023-02-23_203518](https://user-images.githubusercontent.com/3965259/220912856-943db3ca-a509-4e3e-befe-043b85097690.jpg)
![2023-02-23_203535](https://user-images.githubusercontent.com/3965259/220912895-fa4af734-887d-4212-af76-19b78cbf5969.jpg)
2. using ip in tui:
![2023-02-23_203735](https://user-images.githubusercontent.com/3965259/220912936-273ec2fa-83b7-421d-9885-04f43ebb39e6.jpg)
![2023-02-23_203746](https://user-images.githubusercontent.com/3965259/220912951-942adba8-406f-4ded-9de3-91f7cd955f13.jpg)
3. using hostname in cli:
![2023-02-23_204525](https://user-images.githubusercontent.com/3965259/220912973-0546550b-9f80-4f9c-89f4-4a4638c047df.jpg)
![2023-02-23_204532](https://user-images.githubusercontent.com/3965259/220913006-7de17de0-9f7a-4aa9-bf1b-ae40de1e3571.jpg)
4. using ip in cli:
![2023-02-23_204604](https://user-images.githubusercontent.com/3965259/220913043-30480ad9-a8b9-4b74-91b5-1ab234a580b7.jpg)
![2023-02-23_204624](https://user-images.githubusercontent.com/3965259/220913061-fbfc06e9-3613-4f06-954f-e6889917e12b.jpg)

(I have also replaced the `%h` in the ssh config with real IP, which did not work either.)

All these approaches failed, so ... what is the correct way?

## Environment
- OS: Windows 11 version 22H2
- Architecture: Intel x86_64
- Rust version: 1.66.1
- termscp version: 0.11.0
- Protocol used SFTP
- Remote server version and name: Alpine Linux 3.17.1

**Comment by veeso** [comment]
Hi,

patterns like `%h` are currently not supported in ssh2-config, so that won't work for sure atm.

The host syntax is not working, maybe there is an issue with comments after the host definition. I need to check if there is a bug in ssh2-config.

In case that won't work I'll may ask you to provide me with the debug log file generated by termscp.

**Comment by veeso** [comment]
It doesn't look like an issue with ssh2-config.

Could you try to run termscp with `-D` option and provide me with the log file (obfuscated at your discretion). The log file can be found at:

```
$HOME/.config/termscp/termscp.log on Linux/BSD
$HOME/Library/Application Support/termscp/termscp.log on MacOs
FOLDERID_RoamingAppData\termscp\termscp.log on Windows
```

**Comment by namespaceYcZ** [comment]
> It doesn't look like an issue with ssh2-config.
> 
> Could you try to run termscp with `-D` option and provide me with the log file (obfuscated at your discretion)? The log file can be found at:
> 
> ```
> $HOME/.config/termscp/termscp.log on Linux/BSD
> $HOME/Library/Application Support/termscp/termscp.log on MacOs
> FOLDERID_RoamingAppData\termscp\termscp.log on Windows
> ```

I changed my virtual machines and ssh config to the following for convenience:
![2023-02-27_213337](https://user-images.githubusercontent.com/3965259/221580267-b32cba02-d170-48dd-b1f1-e5d33a47b6b5.jpg)

ssh to test-ubuntu:
![2023-02-27_213402](https://user-images.githubusercontent.com/3965259/221580308-c0ce889e-327c-4864-9864-9cf63ea1e68f.jpg)

then run termscp with `-D` option:
![2023-02-27_213440](https://user-images.githubusercontent.com/3965259/221580351-d42a5691-d263-416d-b7aa-b9f91f1d406b.jpg)
![2023-02-27_213502](https://user-images.githubusercontent.com/3965259/221580372-9b043b8a-9963-4f10-ae13-e452c505eb90.jpg)
then I found something wrong in the log file: termscp tried to connect to the `test-ubuntu` with my Windows username, not the `cross` specified in the ssh config file:
![2023-02-27_215256](https://user-images.githubusercontent.com/3965259/221581676-6bc04ebe-3762-4ae9-8401-b12a5203a45b.jpg)
so I tried again with username like this:
![2023-02-27_214126](https://user-images.githubusercontent.com/3965259/221582498-78cb9bd5-e230-40e3-98cd-f8442be35a05.jpg)
![2023-02-27_214150](https://user-images.githubusercontent.com/3965259/221584009-18832e46-d951-467e-8725-27aab7bb7ff0.jpg)
termscp used `cross` correctly this time, but still failed with an unknown error:
![image](https://user-images.githubusercontent.com/3965259/221582608-26c00ae3-ed5c-4df5-a153-209fca17cc18.png)

[here are 2 log files found in `%appdata%/termscp/`](https://gist.github.com/namespaceYcZ/041efa09982d1325638ac82f8ae8c781)

**Comment by namespaceYcZ** [comment]
> It looks like the ssh config resolves correctly and RSA key is actually found, but for some reason, the ssh crate reports an unknown error when authenticating. A last try could be to insert the private ssh key in the termscp key storage (CTRL+C, ssh keys) and see if anything changes, but I don't think so.
> 
> I'll investigate on some issues with RSA auth with ssh crate.

I tried to insert the private ssh key (from `id_rsa` in my `~/.ssh/` directory) in the termscp key storage, but failed.
(

1. press `Ctrl+N` in `SSH keys` tab
2. enter ip address and username
3. termscp open my editor `hx`
4. paste my private key into it, save and quit.
5. termscp said `ssh key is empty`

)
![2023-02-28_205819](https://user-images.githubusercontent.com/3965259/221861400-a3951b2d-142d-4f2c-8b11-e7cbf13ef616.jpg)

But ... termscp did create a ssh key file named `cross@192.168.0.106.key` in `%appdata%/termscp/.ssh`.
So I tried `termscp -D cross@192.168.0.106` ... and failed again:
[termscp.log](https://github.com/veeso/termscp/files/10850594/termscp.log)

**Comment by veeso** [comment]
I don't know hx as editor, maybe it's running on another thread after start. I'd say try with nano or vim

**Comment by namespaceYcZ** [comment]
> I don't know hx as editor, maybe it's running on another thread after start. I'd say try with nano or vim

I tried again and this problem just disappeared. Both `hx` and `nvim` work.
I don't know why but just let's forget this little interlude.
I installed termscp on my ubuntu virtual machine and tried to connect to other linux virtual machine.
I failed again and found a new problem:
![2023-03-01_202526](https://user-images.githubusercontent.com/3965259/222139916-05730b90-8bc0-4bf5-a352-2d0d5c2a0c42.jpg)
(I need the `PubkeyAcceptedKeyTypes` field to bypass the the new version of the ssh client's restriction)
[termscp.log](https://github.com/veeso/termscp/files/10860957/termscp.log)

**Comment by veeso** [comment]
Can you try temporarily to remove that field to see whether termscp works? Unfortunately, that field will be parsed correctly only in the 0.11.1 version.

**Comment by namespaceYcZ** [comment]
> Can you try temporarily to remove that field to see whether termscp works? Unfortunately, that field will be parsed correctly only in the 0.11.1 version.

I tried again.
The good news is that termscp can connect to my test server with specified username(`termscp testuser@prod-rhel`).
The bad news is that termscp fails to connect to the remote server without a specified username if the local username  is not the same as the remote username. It seems that termscp ignores the `User` field in the ssh config file and use local username.

[Video_2023-03-04_092744.webm](https://user-images.githubusercontent.com/3965259/222868577-c567bd64-791e-41cd-be0d-ee4371efa382.webm)
[tscp-failed.log](https://github.com/veeso/termscp/files/10887225/tscp-failed.log)
[tscp-success.log](https://github.com/veeso/termscp/files/10887226/tscp-success.log)

(Termscp logs in Windows are [here](https://github.com/veeso/termscp/issues/149#issuecomment-1446378050). I think this problem may be Windows-specific. )

**Comment by namespaceYcZ** [comment]
> I tried again. The good news is that termscp can connect to my test server with specified username(`termscp testuser@prod-rhel`). The bad news is that termscp fails to connect to the remote server without a specified username if the local username is not the same as the remote username. It seems that termscp ignores the `User` field in the ssh config file and uses local username.

I have a couple of suggestions, although I don't know if they are appropriate. The first is to not show the `Password:` prompt when logging in successfully with the ssh key, just like the ssh tool does. Another is to put the `termscp.log` file in the cache directory instead of the config directory (if the project uses `dirs` crate, `dirs::cache_dir()` can be used , the cache directory is `~/.cache` on Linux and `~\AppData\Local` on Windows)

**Comment by veeso** [comment]
I fixed in 0.12 the issue with the username not being taken from the ssh config.

**Comment by xd-mau5** [comment]
Hi, i just wanna say i'm still having this error with ssh config, even if i put the key into the key storage
![image](https://github.com/veeso/termscp/assets/73971855/bf8f1997-48f0-4a08-a735-910be0366cee)
